src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,343 @@
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+         double eval = metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(targetNeighbors(j, i))) -
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+}
+
+//! Compute gradient over whole dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        arma::mat& gradient)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = 0; i < dataset.n_rows; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      gradient += (1 - regularization) * OuterProduct(cij);
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate gradient due to triplets.
+        arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        arma::vec cil = dataset.col(i) - dataset.col(impostors(l, i));
+
+        gradient += regularization * (OuterProduct(cij) - OuterProduct(cil));
+      }
+    }
+  }
+}
+
+//! Compute gradient over a batch of data points.
+template<typename MetricType>
+template <typename GradType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        const size_t begin,
+                                        GradType& gradient,
+                                        const size_t batchSize)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = begin; i < begin + batchSize; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      gradient += (1 - regularization) * OuterProduct(cij);
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate gradient due to triplets.
+        arma::vec cij = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        arma::vec cil = dataset.col(i) - dataset.col(impostors(l, i));
+
+        gradient += regularization * (OuterProduct(cij) - OuterProduct(cil));
+      }
+    }
+  }
+}
+
+} // namespace lmnn
+} // namespace mlpack
+
+namespace mlpack {
+namespace optimization{
+
+inline void Projection(arma::mat& iterate)
+{
+  // Projection.
+  arma::cx_vec eigval;
+  arma::cx_mat eigvec;
+
+  arma::eig_gen(eigval, eigvec, iterate);
```

Comment:
`iterate` has to be symmetric (since we are optimizing M), so you can use `eig_sym()` here which will make your life a bit easier.